controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -186,29 +208,37 @@ private void triggerReadyJobs(LockedApplication application) {
      */
     private boolean changesAvailable(Application application, JobStatus previous, JobStatus next) {
         if ( ! application.deploying().isPresent()) return false;
-        Change change = application.deploying().get();
 
-        if ( ! previous.lastSuccess().isPresent()) return false;
+        Change change = application.deploying().get();
 
-        if (change instanceof Change.VersionChange) {
+        if (change instanceof Change.VersionChange) { // Propagate upgrade while making sure we never downgrade
             Version targetVersion = ((Change.VersionChange)change).version();
-            if ( ! (targetVersion.equals(previous.lastSuccess().get().version())) )
-                return false; // version is outdated
-            // The below is checked again in allowedTriggering, right before actual triggering.
-            if (next != null && isOnNewerVersionInProductionThan(targetVersion, application, next.type()))
-                return false; // Don't downgrade
-        }
 
-        if (next == null) return true;
-        if ( ! next.lastSuccess().isPresent()) return true;
+            // Is the target version tested?
+            if ( ! lastSuccessfulIs(targetVersion, JobType.stagingTest, application))
+                return false;
+
+            // Is it this job's turn to upgrade?
+            if ( previous.type().isProduction() && ! isOnAtLeastProductionVersion(targetVersion, application, previous.type()))
+                return false;
+
+            // Is the next a test zone which already completed this upgrade successfully?
+            if (next != null && next.type().isTest() && lastSuccessfulIs(targetVersion, next.type(), application))
+                return false;
+
+            // Is the next a production job for a zone already upgraded to this version or a newer one?
+            if (next != null && next.type().isProduction() && isOnAtLeastProductionVersion(targetVersion, application, next.type()))
+                return false;
```

Comment:
This stops retries on production jobs which die after deployment.

Reply:
If they fail then won't be on atLeastProductionversion targetVersion, so how do this fail?

Reply:
If the deployment is successful, but not the rest of the job. We still want to differentiate between the rest of the job failing or not?